packages/completer-extension/src/index.ts

```diff
@@ -45,337 +41,187 @@ namespace CommandIDs {
+    serviceManager.registerProvider(new KernelCompleterProvider());
   }
 };
 
-/**
- * An extension that registers notebooks for code completion.
- */
-const notebooks: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:notebooks',
-  requires: [ICompletionManager, INotebookTracker],
+const manager: JupyterFrontEndPlugin<ICompletionProviderManager> = {
+  id: COMPLETION_MANAGER_PLUGIN,
+  requires: [
+    INotebookTracker,
+    IEditorTracker,
+    IConsoleTracker,
+    ISettingRegistry
+  ],
+  provides: ICompletionProviderManager,
   autoStart: true,
   activate: (
     app: JupyterFrontEnd,
-    manager: ICompletionManager,
-    notebooks: INotebookTracker
-  ): void => {
-    // Create a handler for each notebook that is created.
-    notebooks.widgetAdded.connect((sender, panel) => {
-      const editor = panel.content.activeCell?.editor ?? null;
-      const session = panel.sessionContext.session;
-      // TODO: CompletionConnector assumes editor and session are not null
-      const connector = new CompletionConnector({ session, editor });
-      const handler = manager.register({ connector, editor, parent: panel });
-
-      const updateConnector = () => {
-        const editor = panel.content.activeCell?.editor ?? null;
-        const session = panel.sessionContext.session;
+    notebooks: INotebookTracker,
+    editorTracker: IEditorTracker,
+    consoles: IConsoleTracker,
+    settings: ISettingRegistry
+  ): ICompletionProviderManager => {
+    const AVAILABLE_PROVIDERS = 'availableProviders';
+    const PROVIDER_TIMEOUT = 'providerTimeout';
+    const manager = new CompletionProviderManager();
+    const updateSetting = (settingValues: ISettingRegistry.ISettings): void => {
+      const providersData = settingValues.get(AVAILABLE_PROVIDERS);
+      const timeout = settingValues.get(PROVIDER_TIMEOUT);
+      manager.setTimeout(timeout.composite as number);
+      const selectedProviders = providersData.user || providersData.composite;
+      if (selectedProviders) {
+        const sortedProviders = Object.entries(selectedProviders)
+          .filter(val => val[1] >= 0)
+          .sort(([, rank1], [, rank2]) => rank2 - rank1)
+          .map(item => item[0]);
+        manager.activateProvider(sortedProviders);
+      }
+    };
 
-        handler.editor = editor;
-        // TODO: CompletionConnector assumes editor and session are not null
-        handler.connector = new CompletionConnector({ session, editor });
-      };
+    app.restored.then(() => {
+      const availableProviders = [...manager.getProviders().keys()];
+      settings.transform(COMPLETION_MANAGER_PLUGIN, {
+        fetch: plugin => {
+          const schema = plugin.schema.properties!;
+          availableProviders.forEach((item, index) => {
+            schema[AVAILABLE_PROVIDERS]['properties']![item] = {
+              type: 'number',
+              default: (index + 1) * 100
```

Comment:
The default rank here is computed from the order of the provider's registration, which I suppose to be the same as the order of installation. So users always get their newly installed provider on top of the completer. For the default rank, It would be confusing for the external provider authors to know which rank they should choose because they do not have control over the rank of other providers.